### ISSUE TITLE
check with other collabs (gtmgianni)

### ISSUE THREAD

**Issue by brianbrownton** [brianmorton/cask-automation#6]
Hey @gtmgianni - I was wondering if you use any kind of automated detection for your frequent updates to https://github.com/caskroom/homebrew-cask ?

I noticed you are a frequent contributor and you usually have a collection of updates/PRs when you submit.

Any input would be helpful, thanks.

**Comment by gtmgianni** [comment]
@brianmorton My detection of potentially outdated casks is simply checking to see if their appcast checkpoints have changed, and then flagging those casks to be checked semi-automatically. By "semi-automatic", I mean another script will go through the list of flagged casks and use `curl` to download and display their appcasts, and I will manually check to see if there has been a version bump, before the script invokes `cask-repair` to create a PR, if necessary.

The scripts I use are rather rudimentary, and require a degree of user input, unlike your scripts which seem to be automatic. I can upload my scripts as a gist if you'd like to take a look at them?

**Comment by brianmorton** [comment]
@gtmgianni my script is similar to yours - always requires a degree of human intervention. Instead of using the appcasts (I tried a version using appcasts but couldn't get a successfull signal to noise ratio because many of the appcasts for casks are junk), I probe variations of the version number url to see if there is a new version available.

This seems to work fairly well, but has a couple of limitations:
• some people use difficult-to-guess-the-increments versioning
• some casks get a dev or beta version uploaded which aren't ready for release yet (which the script will find)

Right now I think the script covers just over half of the possible casks (around 2200 of the ~4000 casks).

I would be interested in seeing your gist if you are willing to share, but I don't want you to feel compelled to provide it if you'd rather keep it to yourself!

**Comment by gtmgianni** [comment]
@brianmorton It's all right, I don't mind sharing the scripts.

Here are the two scripts I use (one to make a list of outdated casks, one to iterate through the list and create PRs if necessary):

`checkappcasts.sh`: https://gist.github.com/gtmgianni/da193e66855b99374d87997d6bd3f3c6#file-checkappcasts-sh

`updatecasks.sh`: https://gist.github.com/gtmgianni/12e18abb3e5126f2a479f7273d74b2c7#file-updatecasks-sh

**Comment by brianmorton** [comment]
@gtmgianni glad to see you submitted a PR recently at HBC. I have an improved appcast scanning technique if you're interested in also running it, although it has a few more dependencies than your version:
python 3.6.4
sqlite3

if you're interested, i can help you or you can just have a look. the whole thing runs in about 900-1000 seconds (~15m) which I think is much faster than your bash script.

**Comment by gtmgianni** [comment]
@brianmorton Yeah thanks for that, I wouldn’t mind having a look at your script. Not sure if/when I’ll use it though, as I won’t be contributing much to HBC over the next year due to school commitments and exams.

**Comment by core-code** [comment]
> Instead of using the appcasts (I tried a version using appcasts but couldn't get a
> successfull signal to noise ratio because many of the appcasts for casks are junk),

well, many of the 'appcasts' are really just websites. if we restrict ourselves to checking real appcasts (either SPARKLE RSS or GITHUB ATOM) couldn't we do like 80 percent of them with much less effort?

**Comment by brianmorton** [comment]
I have largely solved those problems, and have spoken with @vitorgalvao about this as well.

I solved the problems by using regex to strip things that are likely to change on websites:
```
    tmp = re.sub("<pubDate>.*</pubDate>","", incText, 0, flags=re.S|re.I)
    tmp = re.sub('<head.*</head>','', tmp, 0, flags=re.S|re.I)
    tmp = re.sub("<script.*</script>","", tmp, 0, flags=re.S|re.I)
```
After that I hash the result and store it in a DB. Next time I scan, I check the new hash with the old hash - if it's different the site has changed.

It's not perfect but it's been good enough for me. Combined with the other ticket you commented on (moving to use `/latest` on github releases), I can get good coverage of anything with an appcast at pretty fast speed (approx 10-20 minutes to scan all casks, depending on your computer and internet connection).

I am moving towards releasing analytics for casks (as homebrew-core currently has), and in that same vein @vitorgalvao and I, along with others, have discussed how to tackle automated scanning. The current thinking is to basically create a meta-set of casks which the automated script will read, and each file will be customized to the cask. Some may be possible to cover with generalities (like github) and just opt out ones that don't conform to the generic way of doing it.

**Comment by core-code** [comment]
> I have largely solved those problems, and have spoken with @vitorgalvao about this as well.

sounds great! so whats missing to move ahead here? because i still frequently find casks that have been outdated for months, even if they have perfectly fine (real) appcasts. just someone to run the script? we have a good connection and can we can do that over here.

where exactly did the discussion take place? i have searched the homebrew-cask issue tracker for related discussions but couldn't find anything.

does that cover casks with no appcast as well? i have put together a proposal to help with any cask that doesn't have a sparkle/github feed here https://github.com/Homebrew/homebrew-cask/issues/52837 but didn't receive any response so far

> The current thinking is to basically create a meta-set of casks which the automated script will read, > and each file will be customized to the cask.

i am not sure i am 100% following how that would look or what that would entail. is the technical discussion online?

**Comment by brianmorton** [comment]
I will try to dig up the issues for you. but my immediate answer for how I was going to solve casks without appcasts is to treat the homepage as the appcast if there is no appcast stanza. I would then slowly try to whittle down the list of casks that fell into that bucket by adding an appcast.

because my scanning method doesn't necessarily require a strict type of feed (I'm ok with appcasts being webpages) this would solve the problem for at least myself ;)

**Comment by brianmorton** [comment]
https://github.com/Homebrew/homebrew-cask/issues/43605

That is a start I think. It was mentioned other places but I can't find it now.

**Comment by vitorgalvao** [comment]
Another relevant discussion: https://github.com/Homebrew/homebrew-cask/issues/41240

**Comment by core-code** [comment]
thanks, i'm reading up on all of the discussion so far.

one tidbit from @vitorgalvao caught my eye

> Casks don’t exactly follow a predictable version pattern you can latch on to. versions are based on
> what’s most convenient for interpolation, not the actual app version. Even if you introduced a system > that could understand every XML structure, that wouldn’t translate to correct cask versions.

thats been a problem for us. cask-versions are essentially download-url-tokens, not app-versions. in most cases they are one and the same. but in many they are more, or something else althogether. 
so we have built a mapping system that transforms the cask-version to the actual version (CFBundleVersion and/or CFShortVersionString) of the app. we have information how to transform from the cask-version to the real app version for about 2300 casks. this is behind the system for non-appcast casks i've proposed (because without knowing which real app-version a cask refers to you couldn't compare it to other app-versions out there). i'd have no problem sharing this data if its useful.
